File: leetcode/148. Sort List/js/sol.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
 var sortList = function(head) {
  if (head == null) {
      return null
  }
  
  let nodeCnt = 0
  let dummy = head
  let arr = []
  while(dummy !== null) {
      arr.push(dummy.val)
      dummy = dummy.next
      nodeCnt += 1
  }
  arr = new Float64Array(arr);
  arr.sort()
  // console.log('nodeCnt: ', nodeCnt, ', arr: ', arr)
  // console.log('nodeCnt: ', nodeCnt, ', arr.sort(): ', arr.sort())
//    let newDummy = new ListNode(-1)
  let lastNode
  let curNode
  for (let it = arr.length - 1; it >= 0; it--) {
      if (it == arr.length - 1) {
          curNode = new ListNode(arr[it], null)
      } else {
          curNode = new ListNode(arr[it], lastNode)
      }
      // lastNode = new ListNode(curNode.val, curNode)
      lastNode = curNode 
      // console.log('curNode: ', curNode, ', lastNode, ', lastNode)
  }
  return curNode
};

File: leetcode/148. Sort List/py3/sol.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

# Merge Sort

class Solution:
    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if (None == head):
            return head
        
        arr = []
        while(head != None):
            arr.append(head.val)
            head = head.next
            
        arr.sort()
#        print('arr: ', arr)
        lastNode = None
        for it in reversed(arr):
            curNode = ListNode(it, lastNode)
            lastNode = curNode
            
        return curNode
    
        
        